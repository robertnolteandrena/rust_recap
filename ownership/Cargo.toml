[package]
name = "ownership"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]

[dev-dependencies]
spectral = "0.6.0"
trybuild = "1.0.85"
