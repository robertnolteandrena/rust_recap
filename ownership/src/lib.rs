pub fn get_string_len(string: String) -> usize {
    string.len()
}
pub fn get_string_len_borrowed(string: &String) -> usize {
    string.len()
}
pub fn get_sin(value: f32) -> f64 {
    value.sin() as f64
}
