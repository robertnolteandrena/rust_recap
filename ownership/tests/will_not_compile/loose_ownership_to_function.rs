use ownership::get_string_len;

fn main() {
    let a = String::from("I am a string");
    let a_len = get_string_len(a);
    //the value a is gone. We do not own it anymore.
    //It is now owned by the function get_string_len
    //Unfortunately that function does not give us the value back.
    //Thus we cannot use it anymore.
    println!("the string {} has lenght {}", a, a_len);
}
