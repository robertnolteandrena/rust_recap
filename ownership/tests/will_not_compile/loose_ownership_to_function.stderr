error[E0382]: borrow of moved value: `a`
  --> tests/will_not_compile/loose_ownership_to_function.rs:10:45
   |
4  |     let a = String::from("I am a string");
   |         - move occurs because `a` has type `String`, which does not implement the `Copy` trait
5  |     let a_len = get_string_len(a);
   |                                - value moved here
...
10 |     println!("the string {} has lenght {}", a, a_len);
   |                                             ^ value borrowed here after move
   |
   = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
   |
5  |     let a_len = get_string_len(a.clone());
   |                                 ++++++++
